Sources/Target/Windows/Thread.cpp

```diff
@@ -81,14 +81,54 @@ void Thread::updateState(DEBUG_EVENT const &de) {
   switch (de.dwDebugEventCode) {
   case EXCEPTION_DEBUG_EVENT:
     _state = kStopped;
-    // We always specify kReasonBreakpoint here, but we should instead analyze
-    // what the ExceptionCode is, and set the stop reason accordingly.
-    _stopInfo.event = StopInfo::kEventStop;
-    _stopInfo.reason = StopInfo::kReasonBreakpoint;
+
     DS2LOG(
         Debug, "exception from inferior, code=%s, address=%p",
         Stringify::ExceptionCode(de.u.Exception.ExceptionRecord.ExceptionCode),
         de.u.Exception.ExceptionRecord.ExceptionAddress);
+
+    switch (de.u.Exception.ExceptionRecord.ExceptionCode) {
+    case EXCEPTION_BREAKPOINT:
+    case EXCEPTION_SINGLE_STEP:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonBreakpoint;
+      break;
+
+    case EXCEPTION_ACCESS_VIOLATION:
+    case EXCEPTION_ARRAY_BOUNDS_EXCEEDED:
+    case EXCEPTION_IN_PAGE_ERROR:
+    case EXCEPTION_STACK_OVERFLOW:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonMemoryError;
+      break;
+
+    case EXCEPTION_FLT_DENORMAL_OPERAND:
+    case EXCEPTION_FLT_DIVIDE_BY_ZERO:
+    case EXCEPTION_FLT_INEXACT_RESULT:
+    case EXCEPTION_FLT_INVALID_OPERATION:
+    case EXCEPTION_FLT_OVERFLOW:
+    case EXCEPTION_FLT_STACK_CHECK:
+    case EXCEPTION_FLT_UNDERFLOW:
+    case EXCEPTION_INT_DIVIDE_BY_ZERO:
+    case EXCEPTION_INT_OVERFLOW:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonMathError;
+      break;
+
+    case EXCEPTION_ILLEGAL_INSTRUCTION:
+    case EXCEPTION_PRIV_INSTRUCTION:
+      _stopInfo.event = StopInfo::kEventStop;
+      _stopInfo.reason = StopInfo::kReasonInstructionError;
+      break;
+
+    case EXCEPTION_INVALID_DISPOSITION:
+    case EXCEPTION_NONCONTINUABLE_EXCEPTION:
+    default:
+      DS2BUG("unsupported exception code: %s",
+             Stringify::ExceptionCode(
+                 de.u.Exception.ExceptionRecord.ExceptionCode));
```

Comment:
I don't know (yet) what `EXCEPTION_INVALID_DISPOSITION` and `
EXCEPTION_NONCONTINUABLE_EXCEPTION` stand for so I was thinking of
implementing support when I hit them for the first time :/

On Wed, Mar 30, 2016 at 12:35 PM, strager <notifications@github.com> wrote:

> In Sources/Target/Windows/Thread.cpp
> <https://github.com/facebook/ds2/pull/240#discussion_r57949492>:
>
> > +      _stopInfo.event = StopInfo::kEventStop;
> > +      _stopInfo.reason = StopInfo::kReasonMathError;
> > +      break;
> > +
> > +    case EXCEPTION_ILLEGAL_INSTRUCTION:
> > +    case EXCEPTION_PRIV_INSTRUCTION:
> > +      _stopInfo.event = StopInfo::kEventStop;
> > +      _stopInfo.reason = StopInfo::kReasonInstructionError;
> > +      break;
> > +
> > +    case EXCEPTION_INVALID_DISPOSITION:
> > +    case EXCEPTION_NONCONTINUABLE_EXCEPTION:
> > +    default:
> > +      DS2BUG("unsupported exception code: %s",
> > +             Stringify::ExceptionCode(
> > +                 de.u.Exception.ExceptionRecord.ExceptionCode));
>
> This'll probably happen a *lot*. You should probably communicate the
> error instead of crashing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ds2/pull/240/files/da279dd327af5b2d20a9796c80817d66e80e2b71..dcee5fbd5253932e250a49640779f0f6c30928a2#r57949492>
>



-- 
Stephane Sezer